### DISCUSSION TITLE
Content vs Delivery

### DISCUSSION THREAD

**Discussion by bdelsey64** [usdot-jpo-ode/jpo-wzdx#37] [General]
I'm part of a team working with MassDOT to implement the WZDx specification on a 3rd party work zone data feed. Although this specification clearly identifies the standards for content, there does not appear to be any requirements nor recommendations with regards to the mechanisms for delivery. For example, should data consumers be expected to continuously poll the feed or should the feed be capable to 'push on change' via subscription? Are these implementation details beyond the scope of this standard or will they eventually be addressed?

**Comment by DeLongR** [comment]
I have spent the last couple days or two getting up to speed on this document and this seems to me to a core issue in this specification. The polling rate directly impacts how this specification can be used, additionally some means of describing the veracity of said data would be very useful (the means of collection). 
Reece DeLong: Signali Traffic

**Comment by barbeau** [comment]
My assumption is that it's a typical REST API with a JSON or XML response that should be polled by clients. It would be good to know if someone is approaching implementations differently.

>The polling rate directly impacts how this specification can be used, 

@DeLongR Could you elaborate on this?

**Comment by sknick-iastate** [comment]
I agree with @barbeau on the polling.

@DeLongR  I completely agree on the veracity of the data.  The [metadata](https://github.com/usdot-jpo-ode/jpo-wzdx/blob/master/full-spec/full-spec.md#metadata) is supposed to contain the location verification method and the location method (where it is identifying the start and end locations) which would describe how the location is collected.  Most other fields would likely be user inputted and will vary from agency to agency.  This is definitely a high priority in Iowa to improve the quality of work zone data and determining the best way to collect this information.

**Comment by bdelsey64** [comment]
Polling is a reasonable mechanism given the relatively static nature of the data as defined in the first version of the specification. However, if some of the new (measured, computed, observed) data points recently suggested by others (e.g. travel time, delay, average speed) are eventually adopted, there will be a valid argument for considering other, more real-time friendly, approaches. Regardless, consumers need to understand the rate at which the producer is updating the various data points so they can determine the suitability of the data for their specific needs and how to most appropriately use that data.

**Comment by DeLongR** [comment]
Sorry for the delay in response. In response to barbeau I think bdelsey64 said it best, regardless of the consumers method, I think the rate of at which the producer updates the data is critical for a user to evaluate the suitability of the data. Embedded systems utilizing this data will explicitly need a defined rate for any real time operations, I would imagine 1 Hz would be in the ballpark. Has anyone compiled a list of consumers and then we could more appropriately define what required data rates might be?

**Comment by orbonjb** [comment]
Great stuff.  The generation of the data certainly seems to be a core aspect of this standard that seems to lacking in the comments.  I'd love to add to this conversation.  In short, existing APIs provided by 3rd party companies already appear accomplish the basic intent of this standard via polled user generated data (e.g. Waze APIs and similar products).  However, having the DOT or similar operating authority directly input data about their own operations would be a welcome addition to this data ecosystem.   

As a potential consumer of this data, I am extremely interested in (1) the cadence of updates, (2) the authenticity of said updates (as noted above) and the (3) accuracy of the information contained within the updates.  My concern is that when implemented in a relevant environment, failure to meet certain thresholds for each of these three categories could invalidate my ability to assign much value in the content and I will obtain equivalent data from other sources.  I'll do my best to elaborate:

**Frequency of Updates:**
The existing standard does provide the option to include metadata, which in turn has a "Datafeed-frequency-update" element that is associated with the cadence of updates. However, the optional nature precludes an external consumer from any knowledge about the regularity of updates.  I would like to offer the proposal to consider making this parameter a required parameter in Common Core Data.   Client applications would ideally like to know how frequency to request a data refresh from the exchange.

**Authenticity of Updates**
Downstream consumers of this data may very well act on the information.  At the most fundamental level, I would be interested in knowing if the data is authentic?  Therefore it seems paramount that some sort of established trust protocol is implemented.   It is unclear how this data will be hosted and what verification protocols are to be employed.  Any elaboration would be welcome.

**Accuracy of Content**
As this is stated to be an exchange, is it envisioned that any sort of feed back mechanisms will be employed to automatically vote out stale data?

Otherwise great stuff.  The content section is really impressive and I could see this being very useful.  Thanks for any consideration.

Jim Orbon
Signali Traffic

**Comment by davidcraig4300** [comment]
Great thread, but my question is more basic.  Where or more specifically, who, should be the server for this data?  Does it make sense for each state to have its own, or should the US have a centralized sever (repository) for these feeds?  Or will this be left for some enterprising 3rd party?  I realize that some users will be very close to real-time and will need a local construction zone feed (DSRC, 4G LTE, or 5G), but I believe the immediate consumers would prefer to poll a centralized server, and the fewer of those, the better.   I know this could immediately be consumed by our company and every other mapping company.  To me the key to this success is the penetration of the standard (how many DOTs and more local gov branches actually use the spec and create the feeds.  Then once the data exists, how easy is it to find.  These are my main concerns.

**Comment by barbeau** [comment]
@davidcraig4300 In Tampa we're currently in the fact-finding phase to figure out how we can publish regional work zone data - this was identified as a regional priority as part of our [Tampa Bay Smart Cities Alliance](https://tbsmartcities.com/). 

This effort currently involves 3 major players - FDOT D7, City of Tampa, and Hillsborough County. Based on preliminary discussions, it looks like each organization is going to need to publish their own WZDx feed. Currently the data is stored in disparate systems (even within each organization), and given IT infrastructure and policies as well as existing vendor/consultant agreements it's far simpler for each organization to just publish what they have in their own feed. Trying to bring together the data from each organization to publish a single WZDx feed would significantly complicate the effort.

This disaggregate approach follows the deployment of General Transit Feed Specification (GTFS) formats, which based on years of experience are typically maintained by the individual transit agencies. This does necessitate additional infrastructure to maintain a directory (and potentially repository) of the feeds to make it efficient for consumers to discover and consume the feeds. For GTFS, [OpenMobilityData](http://openmobilitydata.org) and [TransitLand](https://transit.land/) currently serve those purposes for transit data.

**Comment by arielsgold** [comment]
@davidcraig4300 @barbeau 
During the kickoff for the work zone data working group (WZDWG) on 7/25/19, I said that USDOT plans to aggregate an archive of data feeds from early WZDx publishers. This is to support early adopters and USDOT research, not a long term plan/commitment to aggregate and share these data. We will share the location of that data when it's available and in the meantime will update this repo with links to WZDx feeds when IOOs inform us they are live. If anyone is interested in giving us feedback on the research repository as a beta user, let me know!

Ariel Gold, USDOT, WZDWG Chair

**Comment by davidcraig4300** [comment]
Ariel,

Perfect!  Thank you!  We are very interested in seeing this.

David Craig, Chief of Maps, GM

**Comment by chuehlien** [comment]
Great discussion in this thread! As @arielsgold had mentioned, the USDOT ITS DataHub team is in the process of creating a pipeline to aggregate the WZDx data feeds as a proof of concept. I've started #49 to get feedback on a few upcoming features for consumers of realtime/historical WZDx data, some of them inspired by the discussion in this thread. We would love to get your feedback on #49 and discuss more there!

Julia Lien, USDOT ITS DataHub

**Comment by lynnerandolph** [comment]
I believe that most consumers of this data would prefer to poll rather than subscribe to updates. We have a large number of stakeholders in Texas who receive Center-to-Center data and, in general, most of them prefer access to an XML feed that they can just poll at their desired frequency than to implement the more complicated version of listening to updates.
I do feel that the frequency of the feed update should be a required element, in that case. If I'm a new consumer of the data, I can pull the feed and see that it doesn't get updated more frequently than once an hour or once every 10 minutes and I can base my polling rate on that information.
